### PR TITLE
DEP: scipy.stats.morestats: improve deprecation warnings 

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Begins to address gh-18279

#### What does this implement/fix?
@tupui Thought I'd get your opinion here before opening on the main repo.

This addresses gh-18279 by emitting the right error or warning depending on what attribute is being accessed/imported.

### CODE REVIEW

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -32,11 +33,20 @@ def __dir__():
 def __getattr__(name):
     if name not in __all__:
         raise AttributeError(
-            "scipy.stats.morestats is deprecated and has no attribute "
-            f"{name}. Try looking in scipy.stats instead.")
```
Comment (mdhaber) [inline scipy/stats/morestats.py:36]: AFAICT, if it's not in `__all__`, it's not going to be in `scipy.stats`.

File: scipy/stats/morestats.py [hunk 2]
```diff
@@ -32,11 +33,20 @@ def __dir__():
...
+            "`scipy.stats.morestats` is deprecated and has no attribute "
+            f"`{name}`.")
 
-    warnings.warn(f"Please use `{name}` from the `scipy.stats` namespace, "
-                  "the `scipy.stats.morestats` namespace is deprecated.",
-                  category=DeprecationWarning, stacklevel=2)
+    attr = getattr(scipy.stats, name, None)
```
Comment (mdhaber) [inline scipy/stats/morestats.py:40]: Choose the appropriate message depending on whether the attribute is public in `scipy.stats` or not.

File: scipy/stats/morestats.py [hunk 3]
```diff
@@ -32,11 +33,20 @@ def __dir__():
...
+
+    if attr is not None:
+        message = (f"Please import `{name}` from the `scipy.stats` namespace; "
+                   "the `scipy.stats.morestats` namespace is deprecated and "
+                   "will be removed in SciPy 2.0.0.")
+    else:
+        message = (f"`scipy.stats.morestats.{name}` is deprecated along with "
+                   "the `scipy.stats.morestats` namespace and will be removed "
+                   "in SciPy 1.13.")
```
Comment (mdhaber) [inline scipy/stats/morestats.py]:
```suggestion
                   "the `scipy.stats.morestats` namespace. "
                   f"`scipy.stats.morestats.{name}` will be removed in SciPy 1.13.0, and "
                   "the `scipy.stats.morestats` namespace will be removed in SciPy 2.0.0.")
```

File: scipy/stats/morestats.py [hunk 4]
```diff
@@ -32,11 +33,20 @@ def __dir__():
 def __getattr__(name):
     if name not in __all__:
         raise AttributeError(
-            "scipy.stats.morestats is deprecated and has no attribute "
-            f"{name}. Try looking in scipy.stats instead.")
+            "`scipy.stats.morestats` is deprecated and has no attribute "
+            f"`{name}`.")
```
Comment (mdhaber) [inline scipy/stats/morestats.py]:
```suggestion
            f"`scipy.stats.morestats` has no attribute `{name}`; furthermore, "
            "`scipy.stats.morestats` is deprecated and will be removed in "
            "SciPy 2.0.0.")
```

Comment (tupui) [reply scipy/stats/morestats.py]: "it" could be clarify to say we are talking about the namespace and not the attribute.

**Review by tupui — COMMENTED** [review]
That looks nice. I suppose we can then copy this logic to the other modules then.

**Pull request closed without merging**